### PR TITLE
Sequence packing

### PR DESCRIPTION
**Pull request by fladhak**
This PR contains the changes for sequence packing. This will allow us to have more predictable, consistent batch sizes throughout the training run. We also observed close to 10% improvement in throughput during our testing on H100s.

### CODE REVIEW

**Comment by warner-benjamin** [comment]
This sequence packer is an online dataloader, so it requires scheduling everything in terms of tokens (or maybe batches) and not epochs. This also supports batch size warmup, which can accelerate learning for the same token count.

**Comment by staghado** [comment]
I see that it’s already merged but I wonder if it’s possible to skip spike samples in a deterministic way with packing involved. Thanks.
